arrows/vxl/morphology.cxx

```diff
@@ -0,0 +1,325 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "morphology.h"
+
+#include <arrows/vxl/image_container.h>
+
+#include <vital/util/enum_converter.h>
+
+#include <vital/range/iota.h>
+
+#include <vil/algo/vil_binary_closing.h>
+#include <vil/algo/vil_binary_dilate.h>
+#include <vil/algo/vil_binary_erode.h>
+#include <vil/algo/vil_binary_opening.h>
+#include <vil/algo/vil_structuring_element.h>
+#include <vil/vil_plane.h>
+#include <vil/vil_transform.h>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+namespace {
+
+enum morphology_mode
+{
+  MORPHOLOGY_erode,
+  MORPHOLOGY_dilate,
+  MORPHOLOGY_open,
+  MORPHOLOGY_close,
+  MORPHOLOGY_none,
+};
+
+ENUM_CONVERTER( morphology_converter, morphology_mode,
+                { "erode", MORPHOLOGY_erode }, { "dilate", MORPHOLOGY_dilate },
+                { "open", MORPHOLOGY_open }, { "close", MORPHOLOGY_close },
+                { "none", MORPHOLOGY_none } );
+
+enum element_mode
+{
+  ELEMENT_disk,
+  ELEMENT_jline,
+  ELEMENT_iline,
+};
+
+ENUM_CONVERTER( element_converter, element_mode, { "disk", ELEMENT_disk },
+                { "iline", ELEMENT_iline }, { "jline", ELEMENT_jline } );
+
+enum combine_mode
+{
+  COMBINE_none,
+  COMBINE_union,
+  COMBINE_intersection,
+};
+
+ENUM_CONVERTER( combine_converter, combine_mode, { "none", COMBINE_none },
+                { "union", COMBINE_union },
+                { "intersection", COMBINE_intersection } );
+
+inline bool
+union_functor( bool x1, bool x2 )
+{
+  return x1 || x2;
+}
+
+inline bool
+intersection_functor( bool x1, bool x2 )
+{
+  return x1 && x2;
+}
+
+} // namespace
+
+// ----------------------------------------------------------------------------
+class morphology::priv
+{
+public:
+  priv()
+  {
+  }
+
+  ~priv();
+  // Setup structuring elements
+  void setup_internals();
+  // Compute the morpholigical operation on a single plane of an input image
+  void
+  apply_morphology_to_plane( vil_image_view< bool > const& input,
+                             vil_image_view< bool >& output,
+                             unsigned plane_index );
+  // Perform a morphological operation and optionally combine across channels
+  vil_image_view< bool >
+  perform_morphological_operations( vil_image_view< bool > const& input );
+
+  bool setup{ false };
```

Comment:
"Setup" is easily mistaken for a verb; maybe another name ("is_set_up" or "configured" or some such) would be better?